### ISSUE TITLE
TypeORM MongoDB repository won't save every time

### ISSUE THREAD

**Issue by VierenV** [typeorm/typeorm#2655]
**Issue type:**

[X] bug report

**Database system/driver:**

[X] `mongodb`

**TypeORM version:**

[X] `latest`

When I save an entity, sometimes my data is updated, sometimes not. The exact same request won't work the first time but will the second time. I use save() for updating and inserting. Everything is fine on insertion.

Here is my code :

==> Entity

```
@Entity()
export class User {
  @ObjectIdColumn()
  @Type(() => String)
  id: ObjectID;

  @Column()
  address: string;
}
```
==> DTO

```
export class UpdateUserDto {
  @IsNotEmpty({ message: () => translate('validation.is_not_empty') })
  address: string;
}
```
==> Controller

```
@Put(':userId')
@Authentified()
async update(@Param('userId') userId, @Body() dto: UpdateUserDto) {
  return await this.usersService.update(userId, dto);
}
```
==> usersService :

```
import { MongoRepository } from 'typeorm';
...
private readonly userRepository: MongoRepository<User>
...
async update(userId: string, dto: UpdateUserDto): Promise<User> {
   const user = await this.userRepository.findOneOrFail(userId);
   user.address = dto.address;

   return await this.userRepository.save(user);
}
```

When I find() after the save(), my user address is not updated whereas I received modifiedCount 1 from Mongo. If I repeat the request, this time it is working...

**Comment by dgreene1** [comment]
This used to happen all the time in Mongo if you didn't have journaling set to true.
https://docs.mongodb.com/manual/reference/write-concern/#wc-j

You may or may not want to set `j` to true. There are tradeoffs as mentioned in the link above.

I wish I could help you figure out how to set that in TypeOrm, but I haven't gotten that far yet. I'm currently converting to using it from using Mongoose. If I figure it out, I'll update this ticket.

In the meantime, maybe someone from the TypeOrm team can respond.

**Comment by lonix1** [comment]
@dgreene1 I am also going from mongoose to typeorm, but I've only just started.

I'm concerned that typeorm's support for mongodb isn't "complete" yet - whereas even though mongoose isn't pleasant to use, it's still rock solid.

So having gone through all this already, what has been your experience with typeorm+mongo so far? Can you recommended it?

**Comment by dgreene1** [comment]
@Ionix1 I haven’t used TypeOrm w/ Mongo enough to tell you if it’s wise to switch yet. Personally, these days I always use TypeOrm w/ PostGres because it’s a much more proven path, and also because PostGres has a JsonB type, so I can still work in dynamic types if I choose to. Also PostGres and TypeOrm are much more mated.

Another reason I often choose PostGres over Mongo is because of the exact issue that @VierenV is experiencing which I suspect has nothing to do with TypeOrm and everything to do with the “eventually consistent” nature of MongoDB. So I mostly was posting here to share that turning in journalling will hopefully resolve the issue.

But as for the maturity of TypeOrm’s Mongo features, maybe give it a shot and let us know @Ionix1. But as a warning, the TypeOrm maintainers have shared that the Mongo support is lower priority than the SQL based DBs.

**Comment by VierenV** [comment]
Hey, some feedbacks for you. Since this issue I solved my problem, due to save() mecanism. I got two functions that called save() successively and encountered concurrency problems. The complete explanation : 

In my NodeJS Auth middleware I was logging a timestamp to store the last time user connects to the API. It means I was using save() twice : in the middleware and in my service ! Sometimes it was working because saving the timestamp was quick enough to let my service save the update, sometimes not. Instead of using save() I use updateOne() with $set, no more concurrency problem.

So far I didn't notice other problems using Mongo with TypeORM, but you need to keep in mind save function should be used with "caution". Update every property is safer but less graceful too.

**Comment by dgreene1** [comment]
Yes, it sounds like you created a race condition by saving to the database within a middleware function. It might be wiser to record that time stamp in memory (I believe by adding it to the context object in Koa or the request in Express) and then using the time stamp in your route controller. That way you only write to the database within your controllers or lower. In this way you only update timestamp when a request to the server was successful.

Side note @VierenV you should close this issue since it’s not related to TypeOrm.

**Comment by VierenV** [comment]
For my app we need to log every contact with the API, so it's better for us to gather this update in a middleware, but you're right that we should save only once to avoid database call back and forth.

Closing this issue, thank you !

**Comment by lonix1** [comment]
Thanks for your advice guys!
